packages/field-base/src/input-constraints-mixin.js

```diff
@@ -70,27 +70,32 @@ export const InputConstraintsMixin = dedupingMixin(
       _createConstraintsObserver() {
         // This complex observer needs to be added dynamically instead of using `static get observers()`
         // to make it possible to tweak this behavior in classes that apply this mixin.
-        this._createMethodObserver(`_constraintsChanged(${this.constructor.constraints.join(', ')})`);
+        this._createMethodObserver(`_constraintsChanged(stateTarget, ${this.constructor.constraints.join(', ')})`);
       }
 
       /**
        * Override this method to implement custom validation constraints.
+       * @param {HTMLElement | undefined} stateTarget
        * @param {unknown[]} constraints
        * @protected
        */
-      _constraintsChanged(...constraints) {
-        // Validate the field on constraint change only if it has a value.
-        // The exception is the case when the field is invalid. In that case,
-        // let the method reset `invalid` when the last constraint is removed.
-        if (!this.invalid && !this._hasValue) {
+      _constraintsChanged(stateTarget, ...constraints) {
+        // The input element's validity cannot be determined until
+        // all the necessary constraint attributes aren't set on it.
+        if (!stateTarget) {
           return;
         }
 
-        if (this._hasValidConstraints(constraints)) {
+        const hasConstraints = this._hasValidConstraints(constraints);
+        const isLastConstraintRemoved = this.__previousHasConstraints && !hasConstraints;
+
+        if ((this._hasValue || this.invalid) && hasConstraints) {
```

Comment:
nit: after the change, a field with constraints can't be marked as initially invalid:
`<vaadin-integer-field label="Quantity" min="3" invalid></vaadin-integer-field>`

if the field has no constraints, it works:
`<vaadin-integer-field label="Quantity" invalid></vaadin-integer-field>`

but this is hardly a valid use-case anyway